docs/JSONFormat.md

```diff
@@ -0,0 +1,551 @@
+# YINsolidated Format
+
+The goal of YINsolidated is to generate a single JSON document with all augments and
+submodules merged together. Because the native `yin` format is XML there are some
+oddities when converting to JSON.
+
+What would be a XML node becomes a JSON Object that represents a single `YinElement`
+with the following schema:
+
+- Each object always has a `"keyword": string`. This would be the XML element tag.
+- Sub-objects are grouped under `"children": YinElement[]`.
+- Each object has `namespace: string`
+- Each object has `nsmap: { [prefix: string]: string }` which is a mapping of namespace-prefix to uri
+- All other attributes are optional.
+
+Additionally the following transformations are applied:
+
+## Submodules
+
+All data definitions within a `submodule` are included as child elements of the
+main module element.
+
+## Augments
+
+* All `augment` statements are resolved such that all data definition statements
+  within the `augment` are added as child elements to the element indicated by
+  the `augment`'s target node.
+
+* Any `when` sub-statements of an `augment` are added as child elements to each
+  of the augmenting data definition elements described above. To differentiate
+  these from actual `when` sub-statements of those data definitions, the added
+  `when` elements are given a `context-node="parent"` attribute. This way, the
+  `when` statement still makes the data definition conditional, and the XPath
+  expression can be evaluated against the proper context node.
+
+* Similarly, any `if-feature` sub-statements of an `augment` are added as child
+  elements to each of the augmenting data definitions. No modifications to
+  these elements are necessary because they still make the data definition
+  dependent on the feature.
+
+E.g. this YANG snippet:
+
+```yang
+container augmented-container;
+
+augment "augmented-container" {
+    if-feature my-feature;
+    when "a-leaf != 'foo'";
+
+    container augmenting-container;
+
+    leaf augmenting-leaf {
+        type string;
+    }
+}
+```
+
+Becomes:
+
+```json
+{
+  "keyword": "container",
+  "name": "augmented-container",
+  "children": [
+    {
+      "keyword": "container",
+      "name": "augmenting-container",
+      "children": [
+        { "keyword": "if-feature", "name": "my-feature" },
+        { "keyword": "when", "condition": "a-leaf != 'foo'", "context-node": "parent" }
```

Comment:
I don't think I understand the question. This is copy-pasted and converted from the example in `XMLFormat.md`

Reply:
Okay, I guess if the behavior matches the XML then that's fine but might become an issue that needs to be solved in both places later.

Basically the question is whether these `when` with `context-node=parent` will also trickle down to grandchildren of the augmenting-container, or it if stops at the first level of children. Because I think we have seen issues sometimes where the `when` is applied to the grandchildren, which makes no sense for their evaluation context (the "parent" should really say "grandparent" or something but really the `when` should just not be on those elements).

Does that make more sense?

Reply:
it does. We've seen that with when statements that cross groupings right?

Reply:
Yeah, I think that is typically the less-tested and thornier kind of case where this crops up. I don't know all the details of specific cases but Kaushik ran into some issues where it was propagating `when`s differently than expected. I guess it's something that can be addressed as it comes up.